lib/tree_sequence.c

```diff
@@ -986,6 +1001,300 @@ tree_sequence_simplify(tree_sequence_t *self, node_id_t *samples, size_t num_sam
+    // relationships based upon the tree even if the entries in the mutation
+    // table aren't in order (allowing us to check the order of the table).
+    // We also check and flag various issues related to parentage:
+    //
+    // - If A is declared to be a parent of B, but that relationship is not
+    //   consistent with the node-based parent information in the tree.
+    //
+    // - If A and B are on the same branch, but no parent relationship is
+    //   declared; the parent/child relationship here is determined by order:
+    //   if A occurs before B, then A is the parent of B.
+    //
+    // - If A is the parent of B, but A occurs after B in the mutations table.
+    //   Note that this is an error whether A is explicitly *declared* to be
+    //   the parent of B, or is merely *inferred* to be; however, when A is
+    //   explicitly declared to be the parent of B and they are out of order,
+    //   msprime should catch that issue already on table load anyway.
+    //   
+    // If `fix_missing_parents` is true, then parents currently -1 will be
+    // filled in with the correct value; if it is false, then all parents,
+    // including those set to -1, will be checked for correctness.
+
+    int ret = 0;
+
+    assert(self != NULL);
+
+    if (tree_sequence_get_num_mutations(self) == 0)
+        return 0;
+
+    // start looping through trees
+    sparse_tree_t tree;
+    ret = sparse_tree_alloc(&tree, self, 0);
+    if (ret != 0) goto out;
+    ret = sparse_tree_first(&tree);
+    if (ret < 0) goto out;
+
+    do
+    {
+        // for this tree, loop through sites
+        for (size_t tree_site_index = 0; tree_site_index < tree.sites_length; ++tree_site_index)
+        {
+            site_t *site = &tree.sites[tree_site_index];
+
+            // If there is only one mutation (or none) we can skip the rest of
+            // the work (no parent-child relationship)
+            if (site->mutations_length <= 1)
+                continue;
+
+            // This maps node_id_t to mutation_id_t; think of this as a slow
+            // std::unordered_map<node_id_t, mutation_id_t> we also keep a
+            // couple of auxiliary columns that assist the algorithm; see below
+            static node_map_t *node_map = NULL;
+            static table_size_t node_map__count = 0, node_map__capacity = 0;
+
+            // Ensure the node map has sufficient capacity to hold an entry for
+            // each mutation at this site We expand by doublings, to avoid
+            // repeatedly scootching our size up by small increments
+            if (node_map__capacity < site->mutations_length)
+            {
+                if (node_map__capacity == 0)
+                    node_map__capacity = 16;    // arbitrary start size
+
+                while (node_map__capacity < site->mutations_length)
+                    node_map__capacity <<= 1;
+
+                node_map = realloc(node_map, node_map__capacity * sizeof(node_map_t));
```

Comment:
All mallocs must be checked.  See other ``realloc`` examples elsewhere for the local convention on dealing with reallocs.

Reply:
we're fine with that; the reasoning was that `malloc` never fails (or, if it does, worse things happen) on modern systems; so adding all the checks obfuscates things.

Reply:
Can't say I agree with that reasoning... out of memory conditions occur reasonably often, and you should always try to return a sensible error to the Python interpreter. It's a basic philosophy that I follow: every possible error, ever under any circumstances **must** be accounted for. This is the source of a lot of the weird conventions I've adopted (but I mostly learned them from the CPython code).